File: src/subject_data_analisis/kernels.m
function [decision_kernel,confidence_kernel,decision_kernel_std,confidence_kernel_std] =...
    kernels(tfluct,dfluct,selection,confidence,is_binary_confidence,locked_on_onset,RT_ind)
% Compute psychophysical kernels from target and distractor fluctuations.
% The function allows you to compute them locked to stimulus onset or to
% decision time.
% 
% Sintax:
% [decision_kernel,confidence_kernel,decision_kernel_std,confidence_kernel_std] =...
%      kernels(tfluct,dfluct,selection,confidence,is_binary_confidence,locked_on_onset,RT_ind)
% 
% Input:
%   tfluct: An NxM matrix of target stimulus fluctuations. N is the number
%       of trials and M is the time step index.
%   dfluct: Distractor fluctuations matrix with the same size as tfluct.
%   selection: A binary Nx1 vector indicating if the target (1) or
%       distractor (2) was selected at each trial
%   confidence: Is an Nx1 vector. It can have either a binary values
%       indicating low (1) or high (2) confidence, or be a real value
%       probability of high confidence reports.
%   is_binary_confidence: Boolean indicating if the confidence input is a
%       binary confidence value or not (default true). WARNING! This
%       function is experimental for non binary confidence reports. For
%       continuous confidence reports, it assumes that each value
%       represents the probability of responding high confidence, which can
%       be a wrong interpretation of the data.
%   locked_on_onset: Boolean indicating if the kernels are to be computed
%       locked on stimulus onset (taken to be the first time step). Default
%       if true.
%   RT_ind: Mandatory if locked_on_onset is false. Is an Nx1 vector that
%       indicates the time step at which the subject responded.
% 
% If there are no confidence reports and you want the decision kernels,
% supply confidence = ones(size(selection)) and ignore the confidence
% output.
% 
% outputs:
%    decision_kernel, decision_kernel_std: 2xM vectors containing the
%       selected and non selected patch fluctuations of the decision kernel
%       (and standad deviation).
%    confidence_kernel, confidence_kernel_std: 2xM vectors analogous to the
%       decision_kernel but containg the information on the confidence
%       kernel.
%
% Author: Luciano Paz (July 2015)

    % Default input values
    if nargin<6
        locked_on_onset = true;
        if nargin<5
            is_binary_confidence = true;
        end
    end
    
    if ~locked_on_onset
        % If the locked on onset kernels are wanted, the fluctuations are
        % shifted on a new vector so that RT_ind is located at its center
        
        target_T_dec = nan(size(tfluct,1),2*size(tfluct,2)-1);
        distractor_T_dec = nan(size(dfluct,1),2*size(dfluct,2)-1);
        % Center the luminance fluctuation at response time in the middle of the
        % vector
        for i = 1:size(tfluct,1)
            target_T_dec(i,size(tfluct,2)+1-RT_ind(i):2*size(tfluct,2)-RT_ind(i)) = tfluct(i,:);
            distractor_T_dec(i,size(dfluct,2)+1-RT_ind(i):2*size(dfluct,2)-RT_ind(i)) = dfluct(i,:);
        end
        tfluct = target_T_dec;
        dfluct = distractor_T_dec;
    end
    
    % High and low, selected and non selected fluctuation vectors
    hsel = nan(size(tfluct));
    hnotsel = nan(size(tfluct));
    lsel = nan(size(tfluct));
    lnotsel = nan(size(tfluct));
    hits = selection==1;
    misses = selection==2;
    if is_binary_confidence
        high_confidence = confidence==2;
        low_confidence = confidence==1;
        if any(hits & high_confidence)
            hsel(hits & high_confidence,:) = tfluct(hits & high_confidence,:);
            hnotsel(hits & high_confidence,:) = dfluct(hits & high_confidence,:);
        end
        if any(hits & low_confidence)
            lsel(hits & low_confidence,:) = tfluct(hits & low_confidence,:);
            lnotsel(hits & low_confidence,:) = dfluct(hits & low_confidence,:);
        end
        if any(misses & high_confidence)
            hsel(misses & high_confidence,:) = dfluct(misses & high_confidence,:);
            hnotsel(misses & high_confidence,:) = tfluct(misses & high_confidence,:);
        end
        if any(misses & low_confidence)
            lsel(misses & low_confidence,:) = dfluct(misses & low_confidence,:);
            lnotsel(misses & low_confidence,:) = tfluct(misses & low_confidence,:);
        end
        if ~locked_on_onset
            % Crop the shifted time steps with less than half the trials'
            % information
            hsel(:,sum(~isnan(hsel))<0.5*sum(~all(isnan(hsel),2))) = nan;
            hnotsel(:,sum(~isnan(hnotsel))<0.5*sum(~all(isnan(hnotsel),2))) = nan;
            lsel(:,sum(~isnan(lsel))<0.5*sum(~all(isnan(lsel),2))) = nan;
            lnotsel(:,sum(~isnan(lnotsel))<0.5*sum(~all(isnan(lnotsel),2))) = nan;
        end
        decision_kernel = [nanmean(cat(1,hsel,lsel),1);nanmean(cat(1,hnotsel,lnotsel),1)];
        confidence_kernel = [nanmean(hsel,1)-nanmean(lsel,1);nanmean(hnotsel,1)-nanmean(lnotsel,1)];
        if nargout>3
            decision_kernel_std = [nanstd(cat(1,hsel,lsel),1);nanstd(cat(1,hnotsel,lnotsel),1)]/sqrt(size(tfluct,1));
            confidence_kernel_std = [nanstd(hsel)/sqrt(sum(~all(isnan(hsel),2)))+nanstd(lsel)/sqrt(sum(~all(isnan(lsel),2)));...
                                     nanstd(hnotsel)/sqrt(sum(~all(isnan(hnotsel),2)))+nanstd(lnotsel)/sqrt(sum(~all(isnan(lnotsel),2)))];
        end
    else
        if any(hits)
            hsel(hits,:) = tfluct(hits,:).*repmat(confidence(hits),1,size(tfluct,2));
            lsel(hits,:) = tfluct(hits,:).*repmat((1-confidence(hits)),1,size(tfluct,2));
            hnotsel(hits,:) = dfluct(hits,:).*repmat(confidence(hits),1,size(tfluct,2));
            lnotsel(hits,:) = dfluct(hits,:).*repmat((1-confidence(hits)),1,size(tfluct,2));
        end
        if any(misses)
            hsel(misses,:) = dfluct(misses,:).*repmat(confidence(misses),1,size(tfluct,2));
            lsel(misses,:) = dfluct(misses,:).*repmat((1-confidence(misses)),1,size(tfluct,2));
            hnotsel(misses,:) = tfluct(misses,:).*repmat(confidence(misses),1,size(tfluct,2));
            lnotsel(misses,:) = tfluct(misses,:).*repmat((1-confidence(misses)),1,size(tfluct,2));
        end
        if ~locked_on_onset
            % Crop the shifted time steps with less than half the trials'
            % information
            hsel(:,sum(~isnan(hsel))<0.5*sum(~all(isnan(hsel),2))) = nan;
            hnotsel(:,sum(~isnan(hnotsel))<0.5*sum(~all(isnan(hnotsel),2))) = nan;
            lsel(:,sum(~isnan(lsel))<0.5*sum(~all(isnan(lsel),2))) = nan;
            lnotsel(:,sum(~isnan(lnotsel))<0.5*sum(~all(isnan(lnotsel),2))) = nan;
        end
        decision_kernel = [nanmean(hsel+lsel,1);nanmean(hnotsel+lnotsel,1)];
        confidence_kernel = [nanmean(hsel,1)-nanmean(lsel,1);nanmean(hnotsel,1)-nanmean(lnotsel,1)];
        if nargout>3
            decision_kernel_std = [nanstd(hsel+lsel,1);nanstd(hnotsel+lnotsel,1)]/sqrt(size(tfluct,1));
            confidence_kernel_std = [nanstd(hsel)/sqrt(sum(~all(isnan(hsel),2)))+nanstd(lsel)/sqrt(sum(~all(isnan(lsel),2)));...
                                 nanstd(hnotsel)/sqrt(sum(~all(isnan(hnotsel),2)))+nanstd(lnotsel)/sqrt(sum(~all(isnan(lnotsel),2)))];
        end
    end
end

File: src/subject_data_analisis/kernels.py
#!/usr/bin/python
#-*- coding: UTF-8 -*-
""" Package for computing psychophysical kernels """

import numpy as np
import itertools, math, sys, warnings

def pad_stim(stim,length,axis=1,pad=np.nan,start=0):
	"""
	Takes the stimulus data and returns a numpy.ndarray with an axis
	padded with the desired value up to the desired axis length.
	
	Syntax:
	padded_stim = pad_stim(stim,length,axis=1,pad=np.nan,start=0)
	
	Input:
	- stim: An iterable. Each element must be a numpy.ndarray holding
	  the trial's stimulus. Each element can have different shapes. This
	  function converts all the stimulus data of different shapes into a
	  single ndarray with a shape equal to (Ntrials,...,length,...),
	  where Ntrials is the number of elements in the stim iterable, the
	  ellipses '...' mean that the stim element's shape is copied except
	  for the specified axis that is set to length
	- length: An integer. The length of the padded axis
	- axis: The axis that will be padded and whose length will be set to
	  'length'. Axis cannot be set to 0.
	- pad: The value used to pad the stimulus data
	- start: An integer. The starting index where the stimulus will be
	  copied
	
	Example:
	import numpy as np
	stim = np.random.random((5,2,2))
	print pad_stim(stim,10)
	print pad_stim(stim,10,start=2)
	
	"""
	if axis==0:
		raise(ValueError("axis cannot be set to 0. axis 0 is the trials axis and it cannot be changed"))
	padded_stim = []
	for s in stim:
		shape_s = list(s.shape)
		shape_temp = list(s.shape)
		shape_temp[axis-1] = int(length)
		temp = pad*np.ones(tuple(shape_temp))
		slices = [range(sh) for sh in shape_temp]
		slices[axis-1] = range(start,min(int(length),shape_s[axis-1]+int(start)))
		temp[np.ix_(*slices)] = s.copy()
		padded_stim.append(temp)
	return np.array(padded_stim)

def kernels(fluctuations,selection,confidence,is_binary_confidence=True,locked_on_onset=True,RT_ind=None):
	"""
	Compute psychophysical decision and confidence kernels, locked to
	stimulus onset or to response time.
	
	Syntax: 
	decision_kernel,confidence_kernel,decision_kernel_std,confidence_kernel_std = 
	    kernels(fluctuations,selection,confidence,is_binary_confidence=True,locked_on_onset=True,RT_ind=None)
	
	Input:
	- fluctuations must be a numpy.array or ndarray with shape (Ntrials, Selection index, Time index)
	  with the stimulus fluctuation data for each trial and selectable option
	- selection must be an iterable with Ntrials elements that indicate
	  the selection index for each trial
	- confidence must have the same shape as selection. It can either
	  have binary values indicating high (1) or low confidence (0), or a
	  float indicating the probability of high confidence reports
	- is_binary_confidence is a bool value that flags whether confidence
	  holds binary or real values. Default is true.
	- locked_on_onset is a bool value indicating if the kernels are to
	  be computed locked on stimulus onset (taken to be the first time
	  step). Default is true.
	- RT_ind is mandatory if locked_on_onset is False. Must be an
	  iterable with Ntrials elements that indicate the time step at 
	  which the subject responded.
	
	Output:
	- decision_kernel, decision_kernel_std: 2xM numpy arrays containing
	  the selected and non selected patch fluctuations of the decision
	  kernel (and standad deviation).
	- confidence_kernel, confidence_kernel_std: 2xM numpy arrays
	  analogous to the decision_kernel but containg the information on
	  the confidence kernel.
	"""
	
	fluct_shape = list(fluctuations.shape)
	if not locked_on_onset:
		# If the locked on onset kernels are wanted, the fluctuations are
		# shifted on a new vector so that RT_ind is located at its center
		fluct_shape[2] = fluct_shape[2]*2-1
		fluct_T_dec = np.nan*np.ones(tuple(fluct_shape))
		
		for i,f in enumerate(fluctuations):
			fluct_T_dec[i,:,fluctuations.shape[2]-1-RT_ind[i]:fluct_shape[2]-RT_ind[i]] = f.copy()
		fluctuations = fluct_T_dec
	fluct_shape = tuple(fluct_shape)
	
	# High and low, selected and non selected fluctuation vectors
	high = np.nan*np.ones(fluct_shape)
	low = np.nan*np.ones(fluct_shape)
	high_trials = 0
	low_trials = 0
	for trial,f in enumerate(fluctuations):
		if np.isnan(selection[trial]):
			continue
		if is_binary_confidence:
			if confidence[trial]==1:
				high[trial,0] = f[selection[trial]]
				high[trial,1] = f[1-selection[trial]]
				high_trials+=1
			else:
				low[trial,0] = f[selection[trial]]
				low[trial,1] = f[1-selection[trial]]
				low_trials+=1
		else:
			high[trial,0] = f[selection[trial]]*confidence[trial]
			high[trial,1] = f[1-selection[trial]]*confidence[trial]
			low[trial,0] = f[selection[trial]]*(1.-confidence[trial])
			low[trial,1] = f[1-selection[trial]]*(1.-confidence[trial])
			high_trials+=1
			low_trials+=1
	if not locked_on_onset:
		# Crop the shifted time steps with less than half the trials'
		# information.
		# The horrible logical operation reads as follows:
		# For every time index, if the number of trials with not nan
		# values is less than half the total number of valid trials,
		# then discard all data with that time index by setting its
		# value to nan.
		high[:,:,np.sum(np.logical_not(np.isnan(high[:,0,:])),axis=0)<0.5*high_trials] = np.nan
		low[:,:,np.sum(np.logical_not(np.isnan(low[:,0,:])),axis=0)<0.5*high_trials] = np.nan
	
	with warnings.catch_warnings():
		warnings.simplefilter("ignore", category=RuntimeWarning)
		decision_kernel = np.nanmean(np.concatenate((high,low),axis=0),axis=0)
		confidence_kernel = np.nanmean(high,axis=0)-np.nanmean(low,axis=0)
		
		decision_kernel_std = np.nanstd(np.concatenate((high,low),axis=0),axis=0)/np.sqrt(fluct_shape[0])
		confidence_kernel_std = np.nanstd(high,axis=0)/np.sqrt(high_trials)+np.nanstd(low,axis=0)/np.sqrt(low_trials)
	return decision_kernel,confidence_kernel,decision_kernel_std,confidence_kernel_std

def test(data_dir='/home/luciano/facultad/dropbox_backup_2015_02_03/LuminanceConfidenceKernels/data/controles'):
	"""
	Compute luminance experiment control subjects' decision and
	confidence kernels. Provide data_dir fullpath if the default path is
	not correct
	"""
	try:
		from matplotlib import pyplot as plt
		loaded_plot_libs = True
	except:
		loaded_plot_libs = False
	import data_io as dio
	# Load subject data
	subjects = dio.unique_subjects(data_dir)
	ms = dio.merge_subjects(subjects)
	dat,t,d = ms.load_data()
	# Average the 4 bar luminances and substract the mean luminance
	t = (np.mean(t,axis=2,keepdims=False).T-dat[:,0]).T
	d = np.mean(d,axis=2,keepdims=False)-50
	# Join target and distractor data into a single variable, fluctuations
	fluctuations = np.transpose(np.array([t,d]),(1,0,2))
	# Pad the data with nans to be able to compute kernels locked on response time
	max_rt = max(dat[:,1])
	ISI = 40.
	padding_length = int(math.ceil(max_rt/ISI))
	if max_rt/ISI==padding_length:
		padding_length+=1
	RT_ind = np.floor_divide(dat[:,1],ISI).astype(int)
	RT_ind[RT_ind==padding_length] = padding_length-1
	fluctuations = pad_stim(fluctuations,padding_length,axis=2)
	
	# Compute kernels locked on stimulus onset
	sdk,sck,sdk_std,sck_std = kernels(fluctuations,1-dat[:,2],dat[:,3]-1)
	sT = np.array(range(padding_length),dtype=float)*ISI
	
	# Compute kernels locked on response time
	rdk,rck,rdk_std,rck_std = kernels(fluctuations,1-dat[:,2],dat[:,3]-1,locked_on_onset=False,RT_ind=RT_ind)
	rT = (np.array(range(2*padding_length-1),dtype=float)-padding_length+1)*ISI
	
	if loaded_plot_libs:
		# Plot kernels
		plt.figure(figsize=(13,10))
		# Decision kernel locked on stimulus onset
		plt.subplot(221)
		plt.plot(sT,sdk[0],color='b')
		plt.plot(sT,sdk[1],color='r')
		plt.fill_between(sT,sdk[0]-sdk_std[0],sdk[0]+sdk_std[0],color='b',alpha=0.3,edgecolor=None)
		plt.fill_between(sT,sdk[1]-sdk_std[1],sdk[1]+sdk_std[1],color='r',alpha=0.3,edgecolor=None)
		plt.plot(plt.gca().get_xlim(),[0,0],color='k')
		plt.ylabel('Fluctuation [$cd/m^{2}$]')
		plt.legend(['$D_{S}$','$D_{N}$'])
		plt.title('Locked on onset')
		# Confidence kernel locked on stimulus onset
		ax = plt.subplot(223,sharex=plt.subplot(221),sharey=plt.subplot(221))
		plt.plot(sT,sck[0],color='b')
		plt.plot(sT,sck[1],color='r')
		plt.fill_between(sT,sck[0]-sck_std[0],sck[0]+sck_std[0],color='b',alpha=0.3,edgecolor=None)
		plt.fill_between(sT,sck[1]-sck_std[1],sck[1]+sck_std[1],color='r',alpha=0.3,edgecolor=None)
		plt.plot(plt.gca().get_xlim(),[0,0],color='k')
		plt.xlabel('Time [ms]')
		plt.ylabel('Fluctuation [$cd/m^{2}$]')
		plt.legend(['$C_{S}$','$C_{N}$'])
		# Decision kernel locked on response time
		plt.subplot(222)
		plt.plot(rT,rdk[0],color='b')
		plt.plot(rT,rdk[1],color='r')
		plt.fill_between(rT,rdk[0]-rdk_std[0],rdk[0]+rdk_std[0],color='b',alpha=0.3,edgecolor=None)
		plt.fill_between(rT,rdk[1]-rdk_std[1],rdk[1]+rdk_std[1],color='r',alpha=0.3,edgecolor=None)
		plt.plot(plt.gca().get_xlim(),[0,0],color='k')
		plt.legend(['$D_{S}$','$D_{N}$'])
		plt.title('Locked on response time')
		# Confidence kernel locked on response time
		plt.subplot(224,sharex=plt.subplot(222),sharey=plt.subplot(222))
		plt.plot(rT,rck[0],color='b')
		plt.plot(rT,rck[1],color='r')
		plt.fill_between(rT,rck[0]-rck_std[0],rck[0]+rck_std[0],color='b',alpha=0.3,edgecolor=None)
		plt.fill_between(rT,rck[1]-rck_std[1],rck[1]+rck_std[1],color='r',alpha=0.3,edgecolor=None)
		plt.plot(plt.gca().get_xlim(),[0,0],color='k')
		plt.xlabel('Time - RT [ms]')
		plt.legend(['$C_{S}$','$C_{N}$'])
		plt.show()
	else:
		print sdk,sck,sdk_std,sck_std
		print rdk,rck,rdk_std,rck_std
	return 0

if __name__=="__main__":
	if len(sys.argv)>1:
		test(sys.argv[1])
	else:
		test()
